OSDOCS-2387 MCO supports upgrades-in-place for single-node OpenShift (#37584)
openshift/openshift-docs: modules/updating-sno.adoc

// Module included in the following assemblies:
//
// * updating/updating-cluster-between-minor.adoc

= Updating single node {product-title}

You can update, or upgrade, a single-node {product-title} cluster using either the console or CLI methods. 

However, because single node {product-title} does not support high availability, the following prerequisites are not required: 

> * having a recent etcd backup

Reviewer: @hexfusion so you say that we should encourage customers to perform the etcd backup? +1 from me, just that from what you said, the restore is not supported currently. WDYT?

> * having a recent etcd backup

Reviewer 2: @romfreiman Restore should be supported as long as the existing node has not been destroyed. So you have 2 clear failure cases.

- control-plane node exists restore would be supported in my mind.
- control-plane node does not exist. this is basically a fresh install and not supported.

> * having a recent etcd backup

Author: @hexfusion @romfreiman @sinnykumari @rioliu-rh So, we should not state that the  _having a recent etcd backup_ prerequisite is not required? 
We would just have:
_However, because single-node OpenShift Container Platform does not support high availability, the prerequisite to pause the `MachineHealthCheck` resources is not required._

> * having a recent etcd backup

Author: @hexfusion @romfreiman @sinnykumari @rioliu-rh  ^^

> * having a recent etcd backup

Reviewer 2: makes sense to me I will defer to @romfreiman

* pausing the `MachineHealthCheck` resources

Upgrading single node {product-title} requires downtime and can include an automatic reboot. The amount of downtime depends on the upgrade payload: 

* If the update payload contains an operating system update, which requires a reboot, the downtime is significant and impacts cluster management and user workloads. 

> * If the update has machine configuration changes, but does not require a reboot, the downtime is less and the impact on the cluster management and user workloads is lessened. In this case the nodes are not drained, because there is no other node in the cluster to reschedule the workloads.

Reviewer 3: It will be more clear to mention like this "node drain is not performed on Single Node (or skipped) for changes requiring reboot or a MachineConfig change that would have performed drain on HA cluster, because there is no other node in the cluster to reschedule the workloads"

> * If the update has machine configuration changes, but does not require a reboot, the downtime is less and the impact on the cluster management and user workloads is lessened. In this case the node drain is skipped with single node {product-title} because because there is no other node in the cluster to reschedule the workloads.

Reviewer 3: Minor nit:  two times "because because"

// Module included in the following assemblies:
//
// * updating/updating-cluster-between-minor.adoc
// * updating/updating-cluster-cli.adoc
// * updating/updating-cluster.adoc

> 

Reviewer 4: Please add the anchor.

Also, you shouldn't start a concept's title with a gerund.

= Updating single node {product-title}

> You can update, or upgrade, a single-node {product-title} cluster using either the console or CLI. 

Reviewer 4: ```suggestion
You can update, or upgrade, a single-node {product-title} cluster by using either the web console or CLI. 
```

However, note the following: 

> 

Reviewer 4: You need a noun after following. Limitations, perhaps?

> * The prerequisite to pause the `MachineHealthCheck` resources is not required, because there is no other node to perform the health check.

Reviewer 4: ```suggestion
* The prerequisite to pause the `MachineHealthCheck` resources is not required because there is no other node to perform the health check.
```

> * Restoring a single-node {product-title} cluster using an etcd backup is not officially supported. However, it is good practice to perform the etcd backup in case your upgrade fails. You might be able to restore your cluster to a previous state using the backup if your control plane is healthy.  

Reviewer 4: ```suggestion
* Restoring a single-node {product-title} cluster by using an etcd backup is not officially supported. However, it is good practice to perform the etcd backup in case your upgrade fails. If your control plane is healthy, you might be able to restore your cluster to a previous state by using the backup.  
```

> * Updating a single-node {product-title} cluster requires downtime and can include an automatic reboot. The amount of downtime depends on the update payload: 

Reviewer 4: ```suggestion
* Updating a single-node {product-title} cluster requires downtime and can include an automatic reboot. The amount of downtime depends on the update payload, as described in the following scenarios: 
```

** If the update payload contains an operating system update, which requires a reboot, the downtime is significant and impacts cluster management and user workloads. 

> ** If the update contains machine configuration changes that do not require a reboot, the downtime is less and the impact on the cluster management and user workloads is lessened. In this case, the node drain is skipped with single-node {product-title} because there is no other node in the cluster to reschedule the workloads.

Reviewer 4: ```suggestion
** If the update contains machine configuration changes that do not require a reboot, the downtime is less, and the impact on the cluster management and user workloads is lessened. In this case, the node draining step is skipped with single-node {product-title} because there is no other node in the cluster to reschedule the workloads to.
```

 
** If the update payload does not contain an operating system update or machine configuration changes, a short API outage occurs and resolves quickly. 

[IMPORTANT]
====

> There are conditions, such as bugs in an updated package, that could cause the single node to not restart after a reboot. In this case, the update does not rollback automatically. 

Reviewer 4: ```suggestion
There are conditions, such as bugs in an updated package, that can cause the single node to not restart after a reboot. In this case, the update does not rollback automatically. 
```
Is there anything else helpful we can say here as super-basic troubleshooting or next steps?
